File: kltn-backend/src/main/java/iuh/fit/se/services/impls/CartService.java
package iuh.fit.se.services.impls;

import iuh.fit.se.dtos.requests.AddToCartRequest;
import iuh.fit.se.dtos.requests.UpdateCartItemRequest;
import iuh.fit.se.dtos.responses.CartItemResponse;
import iuh.fit.se.dtos.responses.CartResponse;
import iuh.fit.se.entities.*;
import iuh.fit.se.exceptions.AppException;
import iuh.fit.se.exceptions.ErrorCode;
import iuh.fit.se.mapper.CartItemMapper;
import iuh.fit.se.mapper.CartMapper;
import iuh.fit.se.repositories.*;
import iuh.fit.se.services.interfaces.ICartService;
import iuh.fit.se.services.interfaces.IPriceService;

import java.util.List;
import java.util.Optional;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class CartService implements ICartService {

  CartRepository cartRepository;
  CartItemRepository cartItemRepository;
  UserRepository userRepository;
  ProductRepository productRepository;
  ProductVariantRepository productVariantRepository;
  CartMapper cartMapper;
  CartItemMapper cartItemMapper;
  IPriceService priceService;

  @Override
  public CartResponse getUserCart(Long userId) {
    log.info("Getting cart for user ID: {}", userId);

    User user =
        userRepository
            .findById(userId)
            .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

    Cart cart = cartRepository.findByUserId(userId).orElseGet(() -> createNewCart(user));

    return buildCartResponse(cart);
  }

  @Override
  @Transactional
  public CartItemResponse addToCart(AddToCartRequest request) {
    log.info("Adding product to cart for user ID: {}", request.getUserId());

    // Validate user
    User user =
        userRepository
            .findById(request.getUserId())
            .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

    // Validate product
    Product product =
        productRepository
            .findById(request.getProductId())
            .orElseThrow(() -> new AppException(ErrorCode.PRODUCT_NOT_FOUND));

    // Validate product variant if provided
    ProductVariant productVariant = null;
    if (request.getProductVariantId() != null) {
      productVariant =
          productVariantRepository
              .findById(request.getProductVariantId())
              .orElseThrow(() -> new AppException(ErrorCode.PRODUCT_VARIANT_NOT_FOUND));

      // Check stock availability
      if (productVariant.getStockQuantity() < request.getQuantity()) {
        throw new AppException(ErrorCode.PRODUCT_VARIANT_OUT_OF_STOCK);
      }
    } else {
      // Check base product stock if no variant specified
      if (product.getBasePrice() == null) {
        throw new AppException(ErrorCode.PRODUCT_NOT_FOUND);
      }
    }

    // Get or create cart
    Cart cart =
        cartRepository.findByUserId(request.getUserId()).orElseGet(() -> createNewCart(user));

    // Check if item already exists in cart
    Optional<CartItem> existingItem;
    if (request.getProductVariantId() != null) {
      existingItem =
          cartItemRepository.findByCartIdAndProductIdAndProductVariantId(
              cart.getId(), request.getProductId(), request.getProductVariantId());
    } else {
      existingItem =
          cartItemRepository.findByCartIdAndProductIdAndProductVariantIsNull(
              cart.getId(), request.getProductId());
    }

    CartItem cartItem;
    if (existingItem.isPresent()) {
      // Update quantity
      cartItem = existingItem.get();
      cartItem.setQuantity(cartItem.getQuantity() + request.getQuantity());
    } else {
      // Create new cart item
      cartItem =
          CartItem.builder()
              .quantity(request.getQuantity())
              .cart(cart)
              .product(product)
              .productVariant(productVariant)
              .build();
    }

    CartItem savedItem = cartItemRepository.save(cartItem);
    log.info("Product added to cart successfully");

    return cartItemMapper.toCartItemResponse(savedItem, priceService);
  }

  @Override
  @Transactional
  public CartItemResponse updateCartItem(Long cartItemId, UpdateCartItemRequest request) {
    log.info("Updating cart item with ID: {}", cartItemId);

    CartItem cartItem =
        cartItemRepository
            .findById(cartItemId)
            .orElseThrow(() -> new AppException(ErrorCode.CART_ITEM_NOT_FOUND));

    if (request.getQuantity() <= 0) {
      throw new AppException(ErrorCode.INVALID_QUANTITY);
    }

    cartItem.setQuantity(request.getQuantity());
    CartItem savedItem = cartItemRepository.save(cartItem);

    log.info("Cart item updated successfully");
    return cartItemMapper.toCartItemResponse(savedItem, priceService);
  }

  @Override
  @Transactional
  public void removeCartItem(Long cartItemId) {
    log.info("Removing cart item with ID: {}", cartItemId);

    if (!cartItemRepository.existsById(cartItemId)) {
      throw new AppException(ErrorCode.CART_ITEM_NOT_FOUND);
    }

    cartItemRepository.deleteById(cartItemId);
    log.info("Cart item removed successfully");
  }

  @Override
  @Transactional
  public void clearCart(Long userId) {
    log.info("Clearing cart for user ID: {}", userId);

    Cart cart =
        cartRepository
            .findByUserId(userId)
            .orElseThrow(() -> new AppException(ErrorCode.CART_NOT_FOUND));

    cartItemRepository.deleteByCartId(cart.getId());
    log.info("Cart cleared successfully");
  }

  @Override
  public List<CartItemResponse> getCartItems(Long userId) {
    log.info("Getting cart items for user ID: {}", userId);

    Cart cart =
        cartRepository
            .findByUserId(userId)
            .orElseThrow(() -> new AppException(ErrorCode.CART_NOT_FOUND));

    List<CartItem> cartItems = cartItemRepository.findByCartId(cart.getId());
    return cartItems.stream().map(cartItem -> cartItemMapper.toCartItemResponse(cartItem, priceService)).toList();
  }

  @Override
  public Double getCartTotal(Long userId) {
    log.info("Calculating cart total for user ID: {}", userId);

    Cart cart = cartRepository.findByUserId(userId).orElse(null);
    if (cart == null) {
      return 0.0;
    }

    List<CartItem> cartItems = cartItemRepository.findByCartId(cart.getId());
    return cartItems.stream()
        .mapToDouble(item -> priceService.getFinalPrice(item.getProduct(), item.getProductVariant()) * item.getQuantity())
        .sum();
  }

  @Override
  public Integer getCartItemCount(Long userId) {
    log.info("Getting cart item count for user ID: {}", userId);

    Cart cart = cartRepository.findByUserId(userId).orElse(null);
    if (cart == null) {
      return 0;
    }

    return cartItemRepository.countItemsByCartId(cart.getId()).intValue();
  }

  private Cart createNewCart(User user) {
    Cart cart = Cart.builder().user(user).build();
    return cartRepository.save(cart);
  }

  private CartResponse buildCartResponse(Cart cart) {
    List<CartItem> cartItems = cartItemRepository.findByCartId(cart.getId());

    CartResponse response = cartMapper.toCartResponse(cart);

    // Set cart items
    List<CartItemResponse> cartItemResponses =
        cartItems.stream().map(cartItem -> cartItemMapper.toCartItemResponse(cartItem, priceService)).toList();
    response.setCartItems(cartItemResponses);

    // Calculate totals
    double totalAmount =
        cartItemResponses.stream()
            .mapToDouble(item -> item.getItemTotal() != null ? item.getItemTotal() : 0.0)
            .sum();
    response.setTotalAmount(totalAmount);

    int totalItems =
        cartItemResponses.stream().mapToInt(item -> item.getQuantity().intValue()).sum();
    response.setTotalItems(totalItems);

    return response;
  }
}

File: kltn-frontend/src/services/cartService.ts

import axiosInstance from '../lib/axios';
import { Cart, CartItem, AddToCartPayload, UpdateCartItemPayload } from '@/types/cart';

interface ApiResponse<T> {
  code: number;
  message: string;
  result: T;
}

// Add item to cart
export const addToCart = async (payload: AddToCartPayload): Promise<CartItem> => {
  const response = await axiosInstance.post<ApiResponse<CartItem>>('/cart/add', payload);
  if (response.data.code !== 201) {
    throw new Error(response.data.message || 'Failed to add item to cart');
  }
  return response.data.result;
};

// Get user's cart
export const getCartByUserId = async (userId: number): Promise<Cart> => {
  const response = await axiosInstance.get<ApiResponse<Cart>>(`/cart/user/${userId}`);
  if (response.data.code !== 200) {
    throw new Error(response.data.message || 'Failed to retrieve cart');
  }
  return response.data.result;
};

// Update cart item quantity
export const updateCartItem = async (cartItemId: number, payload: UpdateCartItemPayload): Promise<CartItem> => {
  const response = await axiosInstance.put<ApiResponse<CartItem>>(`/cart/items/${cartItemId}`, payload);
  if (response.data.code !== 200) {
    throw new Error(response.data.message || 'Failed to update cart item');
  }
  return response.data.result;
};

// Remove item from cart
export const removeCartItem = async (cartItemId: number): Promise<void> => {
  const response = await axiosInstance.delete<ApiResponse<null>>(`/cart/items/${cartItemId}`);
  if (response.data.code !== 200 && response.data.code !== 204) { // Handle 204 No Content as success
    throw new Error(response.data.message || 'Failed to remove cart item');
  }
};

// Clear the entire cart
export const clearCart = async (userId: number): Promise<void> => {
  const response = await axiosInstance.delete<ApiResponse<null>>(`/cart/user/${userId}/clear`);
  if (response.data.code !== 200 && response.data.code !== 204) {
    throw new Error(response.data.message || 'Failed to clear cart');
  }
};

// Get total amount of the cart
export const getCartTotal = async (userId: number): Promise<number> => {
  const response = await axiosInstance.get<ApiResponse<number>>(`/cart/user/${userId}/total`);
  if (response.data.code !== 200) {
    throw new Error(response.data.message || 'Failed to get cart total');
  }
  return response.data.result;
};

// Get number of items in the cart
export const getCartItemCount = async (userId: number): Promise<number> => {
  const response = await axiosInstance.get<ApiResponse<number>>(`/cart/user/${userId}/count`);
  if (response.data.code !== 200) {
    throw new Error(response.data.message || 'Failed to get cart item count');
  }
  return response.data.result;
};
